server/grpc_service.go

```diff
@@ -234,8 +236,50 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
 	}, nil
 }
 
+const regionHeartbeatSendTimeout = 5 * time.Second
+
+var errSendRegionHeartbeatTimeout = errors.New("send region heartbeat timeout")
+
+// heartbeatServer wraps PD_RegionHeartbeatServer to ensure when any error
+// occurs on Send() or Recv(), both endpoints will be closed.
+type heartbeatServer struct {
+	stream pdpb.PD_RegionHeartbeatServer
+	closed int32
+}
+
+func (s *heartbeatServer) Send(m *pdpb.RegionHeartbeatResponse) error {
+	if atomic.LoadInt32(&s.closed) == 1 {
+		return io.EOF
+	}
+	done := make(chan error, 1)
+	go func() { done <- s.stream.Send(m) }()
+	select {
+	case err := <-done:
+		if err != nil {
+			atomic.StoreInt32(&s.closed, 1)
+		}
+		return errors.Trace(err)
+	case <-time.After(regionHeartbeatSendTimeout):
+		atomic.StoreInt32(&s.closed, 1)
```

Comment:
if timeout here, the Recv may still be blocked?

Reply:
It's ok to block at `Recv` because each `RegionHeartbeat()` is running in separated goroutine.

Reply:
Maybe a better way is to close the stream, but I find no public API sadly.